pkg/ddl/notifier/subscribe.go

```diff
@@ -0,0 +1,208 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package notifier
+
+import (
+	"context"
+	goerr "errors"
+	"fmt"
+	"time"
+
+	"github.com/pingcap/errors"
+	sess "github.com/pingcap/tidb/pkg/ddl/session"
+	"github.com/pingcap/tidb/pkg/kv"
+	"github.com/pingcap/tidb/pkg/sessionctx"
+	"github.com/pingcap/tidb/pkg/util/logutil"
+	"go.uber.org/zap"
+)
+
+// SchemaChangeHandler function is used by subscribers to handle the
+// SchemaChangeEvent generated by the publisher (DDL module currently). It will
+// be called at least once for every SchemaChange. The sctx has already started a
+// pessimistic transaction and handler should execute exactly once SQL
+// modification logic with it. After the function is returned, subscribing
+// framework will commit the whole transaction with internal flag modification to
+// provide exactly-once delivery. The handler will be called periodically, with
+// no guarantee about the latency between the execution time and
+// SchemaChangeEvent happening time.
+//
+// The handler function must be registered by RegisterHandler before the
+// ddlNotifier is started. If the handler can't immediately serve the handling
+// after registering, it can return nil to tell the ddlNotifier to act like the
+// change has been handled, or return ErrNotReadyRetryLater to hold the change
+// and re-handle later.
+type SchemaChangeHandler func(
+	ctx context.Context,
+	sctx sessionctx.Context,
+	change *SchemaChangeEvent,
+) error
+
+// ErrNotReadyRetryLater should be returned by a registered handler that is not
+// ready to process the events.
+var ErrNotReadyRetryLater = errors.New("not ready, retry later")
+
+// HandlerID is the type of the persistent ID used to register a handler. Every
+// ID occupies a bit in a BIGINT column, so at most we can only have 64 IDs. To
+// avoid duplicate IDs, all IDs should be defined in below declaration.
+type HandlerID int
+
+const (
+	// TestHandlerID is used for testing only.
+	TestHandlerID HandlerID = iota
+)
+
+// RegisterHandler must be called with an exclusive and fixed HandlerID for each
+// handler to register the handler. Illegal ID will panic. RegisterHandler should
+// not be called after the global ddlNotifier is started.
+//
+// RegisterHandler is not concurrency-safe.
+func RegisterHandler(id HandlerID, handler SchemaChangeHandler) {
+	intID := int(id)
+	if intID < 0 || intID >= 64 {
```

Comment:
Would it be a good idea to turn these magic numbers into constants?

Reply:
I can add comments here. They do not have a meaningful name for the constant